### ISSUE TITLE
mkdir -p can fail in chrooted environment

### ISSUE THREAD

**Issue by timkelty** [timkelty/shipit-shared#37]
https://github.com/timkelty/shipit-shared/issues/30#issuecomment-165056217

**Comment by timkelty** [comment]
@tlartaud can you try this branch: https://github.com/timkelty/shipit-shared/tree/fix-create-dirs-chroot

```
npm install git+ssh://git@github.com:timkelty/shipit-shared.git#fix-create-dirs-chroot --save-dev
```

Basically what I did was make sure the process doesn't exit if the `mkdir -p` commands return an error. Instead, a warning message will print out, and the Shipit process will continue (set-permissions was already behaving this way).

So in your case, it should create the dirs for the _source_ of the symlinks fine, and when it tries to create the dirs for the _target_ it won't be able to, but it wont matter: you'll just get a warning printed out, and the deploy will continue.

**Comment by tlartaud** [comment]
Hi,

I'm okay with the idea, its fine. Will try that in a few hours an will let you know if that works for me.
Thanks.

**Comment by timkelty** [comment]
@tlartaud were you able to test this?

**Comment by tlartaud** [comment]
@timkelty Hi,

I've just tried this, sorry for the delay, had a lot of work.
I installed your node for the `fix-create-dirs-chroot`, and I remade the hack mentionned here : https://github.com/timkelty/shipit-shared/issues/32#issuecomment-145037233, and updated shipit-deploy.

Find below the relevant part of the log.

```
Running "shared:create-dirs" task
Creating shared directories on remote.
Running "mkdir -p /project.domain.com/shared/public/app/uploads" on host "project.domain.com".
Running "mkdir -p /project.domain.com/shared/public/app/cache" on host "project.domain.com".
Running "mkdir -p /project.domain.com/shared/public/app/wp-rocket-config" on host "project.domain.com".
Running "mkdir -p $(dirname /project.domain.com/shared/.env)" on host "project.domain.com".
Running "mkdir -p $(dirname /project.domain.com/shared/public/.htaccess)" on host "project.domain.com".
Running "mkdir -p $(dirname /project.domain.com/shared/public/app/advanced-cache.php)" on host "project.domain.com".
Directory created on remote: /project.domain.com/shared/public/app/wp-rocket-config.
Running "mkdir -p /project.domain.com/releases/20151222113435/public/app/wp-rocket-config" on host "project.domain.com".
Directory created on remote: /project.domain.com/shared/public/app/uploads.
Running "mkdir -p /project.domain.com/releases/20151222113435/public/app/uploads" on host "project.domain.com".
Directory created on remote: /project.domain.com/shared/public/app/cache.
Directory created on remote: /project.domain.com/shared/public/.htaccess.
Directory created on remote: /project.domain.com/shared/public/app/advanced-cache.php.
Running "mkdir -p /project.domain.com/releases/20151222113435/public/app/cache" on host "project.domain.com".
Running "mkdir -p $(dirname /project.domain.com/releases/20151222113435/public/.htaccess)" on host "project.domain.com".
Running "mkdir -p $(dirname /project.domain.com/releases/20151222113435/public/app/advanced-cache.php)" on host "project.domain.com".
Directory created on remote: /project.domain.com/shared/.env.
Running "mkdir -p $(dirname /project.domain.com/releases/20151222113435/.env)" on host "project.domain.com".
@project.domain.com-err mkdir: cannot create directory '/project.domain.com/releases/20151222113435/public/app/wp-rocket-config': File exists
Could not create directory on remote: /project.domain.com/releases/20151222113435/public/app/wp-rocket-config.
@project.domain.com-err mkdir: cannot create directory '/project.domain.com/releases/20151222113435/public/app/uploads': File exists
Could not create directory on remote: /project.domain.com/releases/20151222113435/public/app/uploads.
@project.domain.com-err mkdir: cannot create directory '/project.domain.com/releases/20151222113435/public/app/cache': File exists
Could not create directory on remote: /project.domain.com/releases/20151222113435/public/app/cache.

Running "shared:set-permissions" task
Setting permissions on remote.
Running "chmod -R 755 /project.domain.com/shared/public/app/uploads" on host "project.domain.com".
Running "chmod -R 755 /project.domain.com/shared/public/app/cache" on host "project.domain.com".
Running "chmod -R 755 /project.domain.com/shared/public/app/wp-rocket-config" on host "project.domain.com".
Running "chmod -R 755 /project.domain.com/shared/.env" on host "project.domain.com".
Running "chmod -R 755 /project.domain.com/shared/public/.htaccess" on host "project.domain.com".
Running "chmod -R 755 /project.domain.com/shared/public/app/advanced-cache.php" on host "project.domain.com".
Directory created on remote: /project.domain.com/releases/20151222113435/.env.
Directory created on remote: /project.domain.com/releases/20151222113435/public/.htaccess.
Directory created on remote: /project.domain.com/releases/20151222113435/public/app/advanced-cache.php.
@project.domain.com-err chmod: cannot access '/project.domain.com/shared/public/app/advanced-cache.php': No such file or directory
Permissions set on remote.

Error: Command failed: /bin/sh -c ssh deploy_user@project.domain.com "chmod -R 755 /project.domain.com/shared/public/app/advanced-cache.php"
chmod: cannot access '/project.domain.com/shared/public/app/advanced-cache.php': No such file or directory


Running "shared:link:dirs" task
Running "if ( ! [ -h "/project.domain.com/releases/20151222113435/public/app/uploads" ] ); then rm -rf "/project.domain.com/releases/20151222113435/public/app/uploads" 2> /dev/null; ln -s "/var/www/vhosts/thomaslartaud.com/project.domain.com/shared/public/app/uploads" "/project.domain.com/releases/20151222113435/public/app/uploads"; fi" on host "project.domain.com".
Running "if ( ! [ -h "/project.domain.com/releases/20151222113435/public/app/cache" ] ); then rm -rf "/project.domain.com/releases/20151222113435/public/app/cache" 2> /dev/null; ln -s "/var/www/vhosts/thomaslartaud.com/project.domain.com/shared/public/app/cache" "/project.domain.com/releases/20151222113435/public/app/cache"; fi" on host "project.domain.com".
Running "if ( ! [ -h "/project.domain.com/releases/20151222113435/public/app/wp-rocket-config" ] ); then rm -rf "/project.domain.com/releases/20151222113435/public/app/wp-rocket-config" 2> /dev/null; ln -s "/var/www/vhosts/thomaslartaud.com/project.domain.com/shared/public/app/wp-rocket-config" "/project.domain.com/releases/20151222113435/public/app/wp-rocket-config"; fi" on host "project.domain.com".
Shared directories symlinked on remote.
```

Well, it works. However, its creating a lot of red lines. Maybe should you add a log line into the console to explain to the user that the symlinkPath option, if filled, can generate those ouputs, or you probably gonna receive some issues about this.

Anyway, great job for this, thanks.

**Comment by tlartaud** [comment]
@timkelty Hi, can you merge this please ?

**Comment by tlartaud** [comment]
> Note : It looks like the issue with grunt-shipit and orchestrator has been fixed. Events get fired using shipit init :) The hack is no more needed.

**Comment by timkelty** [comment]
Merged in shipit-shared@4.4.0

**Comment by tlartaud** [comment]
@timkelty Thanks a lot !
